base/compiler/ssair/inlining.jl

```diff
@@ -1169,9 +1169,21 @@ function analyze_single_call!(
         end
     end
 
-    # if the signature is fully covered and there is only one applicable method,
+    local mostly_covered = true
+    if typeof(signature_union) == DataType && typeof(atype) == DataType
+        sig_union_ts = datatype_fieldtypes(signature_union)
+        atype_ts = datatype_fieldtypes(atype)
+        for (idx, atype_t) in enumerate(atype_ts)
+            sigu_t = sig_union_ts[idx]
+            sub = (atype_t <: sigu_t) || atype_t == Any
+            mostly_covered = mostly_covered && sub
+        end
+    end
+    mostly_covered = mostly_covered && !(atype <: signature_union)
```

Comment:
So the purpose of `mostly_covered` seems to special case `x::Any` argument ?
We may want to a more general solution here, since there are arbitrary abstract type other than `Any`.

I wonder if, when there is only a single match, we can eliminate `atype <: signature_union` restriction and rather set `fully_covered = atype <: match.spec_types` ?

Reply:
I tried this approach, but didn't seem to work (giving error in this test: 

```
f_convert_missing(x) = convert(Int64, x)
let ci = code_typed(f_convert_missing, Tuple{Union{Int64, Missing}})[1][1],
    ci_unopt = code_typed(f_convert_missing, Tuple{Union{Int64, Missing}}; optimize=false)[1][1]
    # [snip...]
    @test length(ci.code) >
        length(ci_unopt.code)
end
```

as well as bypassing some MethodErrors here like we talked about : 

```
f20704(::Int) = 1
Base.@pure b20704(@nospecialize(x)) = f20704(x)
@test b20704(42) === 1
@test_throws MethodError b20704(42.0)
```

I have to work on other stuff today; I will come back to this, but setting as draft for now

Reply:
Yeah, it seems like we still need to calculate `fully_covered` correctly when there are multiple matches.